pkg/platform/osmano/osmclient/osm-app.go

```diff
@@ -0,0 +1,287 @@
+	client, err := s.GetClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := client.GetOKAPackage(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	out, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read response body, %s", err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("invalid response reading OKA apps (%d), %s", resp.StatusCode, string(out))
+	}
+
+	apps := []OKAApp{}
+	err = json.Unmarshal(out, &apps)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal OKA apps output, %s", err)
+	}
+	for _, app := range apps {
+		if app.Name == name {
+			return &app, nil
+		}
+	}
+	return nil, nil
+}
+
+func createAppArchive(app *edgeproto.App) ([]byte, error) {
+	// The tar.gz file should contain an archive of the following format:
+	// manifests/app.yml
+	// templates/app-ks.yml
+	buf := bytes.Buffer{}
+	gw := gzip.NewWriter(&buf)
+	defer gw.Close()
+	tw := tar.NewWriter(gw)
+	defer tw.Close()
+
+	name := getOSMAppName(app)
+	now := time.Now()
+
+	// write manifest
+	manifestHdr := tar.Header{
+		Name:     "manifests/" + name + ".yml",
+		Size:     int64(len(app.DeploymentManifest)),
+		Mode:     0644,
+		ModTime:  now,
+		Typeflag: tar.TypeReg, // file
+	}
+	if err := tw.WriteHeader(&manifestHdr); err != nil {
+		return nil, fmt.Errorf("failed to write manifest header %v, %s", manifestHdr, err)
+	}
+	num, err := tw.Write([]byte(app.DeploymentManifest))
+	if err != nil {
+		return nil, fmt.Errorf("failed to write manifest to archive, %s", err)
+	}
+	if num != len(app.DeploymentManifest) {
+		return nil, fmt.Errorf("failed to write manifest, only wrote %d of %d bytes", num, len(app.DeploymentManifest))
+	}
+
+	// write template
+	args := appTArgs{
+		Name:      name,
+		Namespace: "default",
+	}
+	tempBuf := bytes.Buffer{}
+	err = appT.Execute(&tempBuf, &args)
+	if err != nil {
+		return nil, fmt.Errorf("failed to execute app template, %s, %s", args, err)
+	}
+	dat := tempBuf.Bytes()
+	templateHdr := tar.Header{
+		Name:     "templates/" + name + "-ks.yml",
+		Size:     int64(len(dat)),
+		Mode:     0644,
+		ModTime:  now,
+		Typeflag: tar.TypeReg,
+	}
+	if err := tw.WriteHeader(&templateHdr); err != nil {
+		return nil, fmt.Errorf("failed to write template header %v, %s", templateHdr, err)
+	}
+	num, err = tw.Write(dat)
+	if err != nil {
+		return nil, fmt.Errorf("failed to write template file to archive, %s", err)
+	}
+	if num != len(dat) {
+		return nil, fmt.Errorf("failed to write template, only write %d of %d bytes", num, len(dat))
+	}
+
+	tw.Close()
+	gw.Close()
+	return buf.Bytes(), nil
+}
+
+type appTArgs struct {
+	Name      string
+	Namespace string
+}
+
+var appT = template.Must(template.New("appT").Parse(appTemplate))
+
+var appTemplate = `apiVersion: v1
+kind: Namespace
+metadata:
+  name: {{ .Namespace }}
+---
+apiVersion: kustomize.toolkit.fluxcd.io/v1
+kind: Kustomization
```

Comment:
Is this basically a flux app definition, or some CRD they wrote themselves?

Reply:
This is a FluxCD CRD, but the parameterization is something specific to OSM. I don't understand all the details, but it looks like the OSM APIs trigger ArgoCD workflows to run their shell scripts to process the data, commit that manifests to their internal git repo, and then push a FluxCD kustomization CRD to the target cluster. FluxCD then runs in the target cluster, pulling from the git repo, and applying the changes (actually deploying the AppInst).